sympy/polys/euclidtools.py

```diff
@@ -1248,6 +1247,386 @@ def dup_zz_heu_gcd(f, g, K):
     raise HeuristicGCDFailed('no luck')
 
 
+def free_symbols_sparse(p):
+    """
+    Takes a sparse matrix p as input and calculates the sum of exponents
+    for each column. It then returns a set of indices where the sum of
+    exponents is non-zero.
+    """
+    exponents = list(map(sum, zip(*p)))
+    return {n for n, e in enumerate(exponents) if e}
```

Comment:
It might be possible here to just look at the first one or two terms to see that all of the generators are in the polynomial. In that case it would be inefficient to loop over every term (there could be many terms).